From 5dbe34063dc63dc2c1465e55df0a6fded4feebe7 Mon Sep 17 00:00:00 2001
From: Flopster101 <nahuelgomez329@gmail.com>
Date: Sun, 8 Mar 2026 10:51:44 -0300
Subject: [PATCH] Allow auto-enabling SBC HD

When persist.bluetooth.sbc_hd.enabled is set to true, SBC will
automatically configure itself with:
- Priority above AAC (disables AAC to ensure SBC selection)
- Dual Channel mode for higher bitrate
- codec_specific_1=0x1337 to trigger HD bitrates in the encoder

The Audio HAL is also updated to map Dual Channel mode to stereo
for proper audio routing.

Enable with: adb shell setprop persist.bluetooth.sbc_hd.enabled true

Change-Id: Ie5ab5526047e144a8197a66349c746b97f8ce868
Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
---
 .../bluetooth/a2dp/A2dpCodecConfig.java       | 19 +++++++++++++++----
 .../aidl/a2dp/codec_status_aidl.cc            |  3 +++
 system/stack/a2dp/a2dp_sbc.cc                 | 14 ++++++++++++++
 3 files changed, 32 insertions(+), 4 deletions(-)

diff --git a/android/app/src/com/android/bluetooth/a2dp/A2dpCodecConfig.java b/android/app/src/com/android/bluetooth/a2dp/A2dpCodecConfig.java
index e06fd19..8f927cf 100644
--- a/android/app/src/com/android/bluetooth/a2dp/A2dpCodecConfig.java
+++ b/android/app/src/com/android/bluetooth/a2dp/A2dpCodecConfig.java
@@ -40,6 +40,7 @@ import java.util.Objects;
  */
 class A2dpCodecConfig {
     private static final String TAG = A2dpCodecConfig.class.getSimpleName();
+    private static final String SBC_HD_ENABLED_PROPERTY = "persist.bluetooth.sbc_hd.enabled";
 
     private final Context mContext;
     private final A2dpNativeInterface mA2dpNativeInterface;
@@ -236,6 +237,8 @@ class A2dpCodecConfig {
             mA2dpSourceCodecPrioritySbc = value;
         }
 
+        boolean sbcHdEnabled = SystemProperties.getBoolean(SBC_HD_ENABLED_PROPERTY, false);
+
         try {
             value =
                     SystemProperties.getInt(
@@ -244,7 +247,10 @@ class A2dpCodecConfig {
         } catch (NotFoundException e) {
             value = BluetoothCodecConfig.CODEC_PRIORITY_DEFAULT;
         }
-        if ((value >= BluetoothCodecConfig.CODEC_PRIORITY_DISABLED)
+        if (sbcHdEnabled) {
+            mA2dpSourceCodecPriorityAac = BluetoothCodecConfig.CODEC_PRIORITY_DISABLED;
+            Log.i(TAG, "SBC HD enabled: AAC codec disabled");
+        } else if ((value >= BluetoothCodecConfig.CODEC_PRIORITY_DISABLED)
                 && (value < BluetoothCodecConfig.CODEC_PRIORITY_HIGHEST)) {
             mA2dpSourceCodecPriorityAac = value;
         }
@@ -300,11 +306,16 @@ class A2dpCodecConfig {
 
         BluetoothCodecConfig codecConfig;
         BluetoothCodecConfig[] codecConfigArray = new BluetoothCodecConfig[6];
-        codecConfig =
+        BluetoothCodecConfig.Builder sbcBuilder =
                 new BluetoothCodecConfig.Builder()
                         .setCodecType(BluetoothCodecConfig.SOURCE_CODEC_TYPE_SBC)
-                        .setCodecPriority(mA2dpSourceCodecPrioritySbc)
-                        .build();
+                        .setCodecPriority(sbcHdEnabled ? 2500 : mA2dpSourceCodecPrioritySbc);
+        if (sbcHdEnabled) {
+            sbcBuilder.setChannelMode(BluetoothCodecConfig.CHANNEL_MODE_DUAL_CHANNEL);
+            sbcBuilder.setCodecSpecific1(0x1337);
+            Log.i(TAG, "SBC HD auto-enabled: forcing Dual Channel with high bitrate profile");
+        }
+        codecConfig = sbcBuilder.build();
         codecConfigArray[0] = codecConfig;
         codecConfig =
                 new BluetoothCodecConfig.Builder()
diff --git a/system/audio_hal_interface/aidl/a2dp/codec_status_aidl.cc b/system/audio_hal_interface/aidl/a2dp/codec_status_aidl.cc
index c4a1973..7a8f547 100644
--- a/system/audio_hal_interface/aidl/a2dp/codec_status_aidl.cc
+++ b/system/audio_hal_interface/aidl/a2dp/codec_status_aidl.cc
@@ -105,6 +105,9 @@ static ChannelMode A2dpCodecToHalChannelMode(const btav_a2dp_codec_config_t& a2d
       return ChannelMode::MONO;
     case BTAV_A2DP_CODEC_CHANNEL_MODE_STEREO:
       return ChannelMode::STEREO;
+    case BTAV_A2DP_CODEC_CHANNEL_MODE_DUAL_CHANNEL:
+      // Dual Channel is stereo from the audio HAL perspective.
+      return ChannelMode::STEREO;
     default:
       return ChannelMode::UNKNOWN;
   }
diff --git a/system/stack/a2dp/a2dp_sbc.cc b/system/stack/a2dp/a2dp_sbc.cc
index c6e2c39..e1d998e 100644
--- a/system/stack/a2dp/a2dp_sbc.cc
+++ b/system/stack/a2dp/a2dp_sbc.cc
@@ -47,6 +47,7 @@
 #include "gd/common/utils.h"
 #include "hardware/bt_av.h"
 #include "internal_include/bt_trace.h"
+#include "osi/include/properties.h"
 #include "stack/include/bt_hdr.h"
 
 #define A2DP_SBC_MAX_BITPOOL 53
@@ -985,6 +986,14 @@ static bool select_best_channel_mode(uint8_t ch_mode, tA2DP_SBC_CIE* p_result,
 static bool select_audio_channel_mode(const btav_a2dp_codec_config_t* p_codec_audio_config,
                                       uint8_t ch_mode, tA2DP_SBC_CIE* p_result,
                                       btav_a2dp_codec_config_t* p_codec_config) {
+  // SBC HD auto-enable: force Dual Channel when the property is enabled.
+  if (osi_property_get_bool("persist.bluetooth.sbc_hd.enabled", false)) {
+    p_result->ch_mode = A2DP_SBC_IE_CH_MD_DUAL;
+    p_codec_config->channel_mode = BTAV_A2DP_CODEC_CHANNEL_MODE_DUAL_CHANNEL;
+    log::info("SBC HD: forcing Dual Channel mode via property");
+    return true;
+  }
+
   switch (p_codec_audio_config->channel_mode) {
     case BTAV_A2DP_CODEC_CHANNEL_MODE_MONO:
       if (ch_mode & A2DP_SBC_IE_CH_MD_MONO) {
@@ -1349,6 +1358,11 @@ tA2DP_STATUS A2dpCodecConfigSbcBase::setCodecConfig(const uint8_t* p_peer_codec_
   if (codec_user_config_.codec_specific_1 != 0) {
     codec_config_.codec_specific_1 = codec_user_config_.codec_specific_1;
   }
+  // SBC HD: force codec_specific_1 = 0x1337 to enable HD bitrate profile.
+  if (osi_property_get_bool("persist.bluetooth.sbc_hd.enabled", false)) {
+    codec_config_.codec_specific_1 = 0x1337;
+    log::info("SBC HD: forcing codec_specific_1=0x1337 for HD bitrate profile");
+  }
   if (codec_user_config_.codec_specific_2 != 0) {
     codec_config_.codec_specific_2 = codec_user_config_.codec_specific_2;
   }
-- 
2.53.0

